Introducing rocksDb checkpoint for db snapshot

nkumar04 opened the pull request:
List of changes in this PR:
- Added api in rocks db client to create, list and remove db checkpoints
- Added db checkpoint metadata to store information in persistence
- Added db checkpoint manager to manage creation, deletion of snapshots
- Added basic gtest to verify the db can be used to get/set keys after creating snapshots
- Added apollo test to verify db checkpoint functinality

itsmeakashgoyal commented on bftengine/src/bftengine/PersistentStorageImp.cpp:
```diff
@@ -964,5 +966,21 @@ bool PersistentStorageImp::getNewEpochFlag() {
+  metadataStorage_->read(DB_CHECKPOINT_DESCRIPTOR, maxBuffSz, outBufPtr, outActualObjectSize);
+  if (!outActualObjectSize)  // Parameter not found
+    return std::nullopt;
+  std::vector<uint8_t> v(outActualObjectSize, 0);
```
make variable name readable.

itsmeakashgoyal reviewed (commented):
Minor comments else LGTM.

pkthapa commented on bftengine/include/bftengine/DbCheckpointManager.hpp:
```diff
@@ -0,0 +1,55 @@
+class DbCheckpointManager {
+ public:
+  void sendInternalCreateDbCheckpointMsg(const SeqNum& seqNum, bool noop);
+  void enableDbCheckpoint(bool enable) { enableDbCheckpoint_ = enable; }
```
FMPOV: It is better to make parameters as `const &` for class objects, structures, string, and other user-defined types only. Primitive datatypes should be always passed as `const`. Reason: It is not so useful to access the primitive types via references because it involves extra work to read from the earlier addresses. Whereas, for user-defined objects, passing `const &` saves stack. Wdyt?

toly-kournik commented:
@nkumar04, 
why IDBCheckpointManager interface implementations (rocks, etc) are in the kvbc folder?
As far as I can see there's no notion of blockchain in the db checkpointing, so it should be part of the basic bft engine functionality

toly-kournik commented on bftengine/include/bftengine/DbCheckpointManager.hpp:
```diff
@@ -0,0 +1,52 @@
+ private:
+  DbCheckpointManager() = default;
+  std::function<void(SeqNum)> createDbChecheckpointCb_;
+  std::function<void(SeqNum)> onSatbleCheckpointCb_;
```
It's fine, but I think we'd better add 
`concord::util::CallbackRegistry<uint64_t> onStableSeqNumCallbacks_;`
in a manner similar to 
`concord::util::CallbackRegistry<uint64_t> onViewNumCallbacks_;` to the `ReplicaImp`

nkumar04 replied:
Fixed

toly-kournik commented on bftengine/include/bftengine/DbCheckpointMetadata.hpp:
```diff
@@ -0,0 +1,62 @@
+using CheckpointId = uint64_t;
+using TimeDuration = std::chrono::duration<long>;
+
+struct DbCheckpointMetadata : public concord::serialize::SerializableFactory<DbCheckpointMetadata> {
```
Do we really need to persist this to a mdt storage?
Can't we derive the checkpoints from the checkpoints directory of database itself?

nkumar04 replied:
There can be a case, when replica crashes while creating db checkpoint. If we find, a checkpoint directory on filesystem but no matching entry in dbCheckpointMetadata, then its probably an in-consistent db checkpoint, so we remove it. Also, we store last checkpoint creation time, for creation of checkpoint after some configured duration

toly-kournik commented on kvbc/include/db_checkpoint.h:
```diff
@@ -0,0 +1,142 @@
+using Status = concordUtils::Status;
+using SystemClock = std::chrono::system_clock;
+using Seconds = std::chrono::seconds;
+class IDbCheckPointManager {
```
As mentioned earlier, the name of this interface can be confused with `DBCheckPointManager` itself, 
and shouldn't be defined in `kvbc`

toly-kournik commented on kvbc/include/Replica.h:
```diff
@@ -228,6 +229,7 @@ class Replica : public IReplica,
   concord::util::ThreadPool blocksIOWorkersPool_;
   std::unique_ptr<concord::client::reconfiguration::ClientReconfigurationEngine> creEngine_;
   std::shared_ptr<concord::client::reconfiguration::IStateClient> creClient_;
+  std::unique_ptr<concord::kvbc::IDbCheckPointManager> dbCheckpointMgr_;
```
I think it's better to have an instance of the `concord::kvbc::IDbCheckPointManager` inside the `DBCheckpointManager`. `dbCheckpointMgr_` has no direct use by `ReplicaImp`

nkumar04 replied:
Fixed

toly-kournik commented on bftengine/include/bftengine/DbCheckpointManager.hpp:
```diff
@@ -0,0 +1,52 @@
+  std::chrono::seconds getLastCheckpointCreationTime() const { return lastCheckpointCreationTime_; }
+
+ public:
+  static DbCheckpointManager& Instance() {
```
nit: lowercase `instance()`

nkumar04 replied:
fixed

toly-kournik commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -4283,6 +4286,18 @@ ReplicaImp::ReplicaImp(bool firstTime,
+    if (getReplicaConfig().maxNumberOfDbCheckpoints && seqNum &&
+        !(seqNum % getReplicaConfig().dbCheckPointWindowSize) &&
+        (timeSinceLastSnapshot >= getReplicaConfig().dbSnapshotIntervalSeconds.count())) {
+      DbCheckpointManager::Instance().onCreateDbCheckpoint(seqNum);
```
It's essentially no `onCreateDbCheckpoint()`, but `createDbCheckpoint()`

nkumar04 replied:
fixed

toly-kournik commented on storage/include/rocksdb/native_client.h:
```diff
@@ -163,6 +163,11 @@ class NativeClient : public std::enable_shared_from_this<NativeClient> {
   ::rocksdb::ColumnFamilyHandle *defaultColumnFamilyHandle() const;
   ::rocksdb::ColumnFamilyHandle *columnFamilyHandle(const std::string &cFamily) const;
 
+  void createCheckpointNative(const uint64_t &checkPointId);
```
Why do we need to define this functionality in `NativeClient` is all it does is calling the `IDBClient` interface?

nkumar04 replied:
This is added only for gtest

nkumar04 commented:
> @nkumar04, why IDBCheckpointManager interface implementations (rocks, etc) are in the kvbc folder? As far as I can see there's no notion of blockchain in the db checkpointing, so it should be part of the basic bft engine functionality

@toly-kournik , that's a good point. I have moved it out of kvbc and also renamed the class. Please let me know if you have any other comments. Thanks!

toly-kournik commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -3402,6 +3403,8 @@ void ReplicaImp::onSeqNumIsStable(SeqNum newStableSeqNum, bool hasStateInformati
     IControlHandler::instance()->onStableCheckpoint();
     ControlStateManager::instance().wedge();
   }
+  auto ret = std::async(std::launch::async,
```
The way the callback is called (`std::async`) is implementation dependent, so only `DbCheckpointManager::instance().onStableCheckPoint(newStableSeqNum);` should be called.
Also sorry for being not very clear about a callback registry.
I suggested adding `concord::util::CallbackRegistry<uint64_t> onSeqNumStable_;` along with  `concord::util::CallbackRegistry<uint64_t> onViewNumCallbacks_;` to `ReplicaImp`and not inside the `DbCheckpointManager`
Personally I don't see much value in having `DbCheckpointManager::addOnStableSeqNum()`, since it basically is not a callback, but a known to  `DbCheckpointManager` function. And adding it in `ReplicaImp::ReplicaImp()` doesn't make much sense.

nkumar04 replied:
Fixed

toly-kournik commented on bftengine/include/bftengine/DbCheckpointManager.hpp:
```diff
@@ -0,0 +1,55 @@
+                                     const std::function<uint64_t()>& getLastBlockIdCb);
+
+ public:
+  static DbCheckpointManager& instance() {
```
Why do we basically need it to be a singleton? I think it's initialized used only from a ReplicaImp

toly-kournik commented on bftengine/src/bftengine/ReplicaImp.hpp:
```diff
@@ -682,6 +682,7 @@ class ReplicaImp : public InternalReplicaApi, public ReplicaForStateTransfer {
   batchingLogic::RequestsBatchingLogic reqBatchingLogic_;
   ReplicaStatusHandlers replStatusHandlers_;
   concord::util::CallbackRegistry<uint64_t> onViewNumCallbacks_;
+  concord::util::CallbackRegistry<SeqNum> createDbCheckpointCb_;
```
@nkumar04, my comment was to introduce
`concord::util::CallbackRegistry<SeqNum> onSeqNumIsStableCallbacks_;`
and eventually to move all the callbacks from `ReplicaImp::onSeqNumIsStable()` to it.
And we don't need a `createDbCheckpointCb_`.

nkumar04 replied:
fixed

toly-kournik commented on bftengine/src/bftengine/ReplicaImp.cpp:
```diff
@@ -3402,6 +3403,8 @@ void ReplicaImp::onSeqNumIsStable(SeqNum newStableSeqNum, bool hasStateInformati
     ControlStateManager::instance().wedge();
   }
+  auto ret =
+      std::async(std::launch::async, [this, newStableSeqNum]() { createDbCheckpointCb_.invokeAll(newStableSeqNum); });
```
the async way of an invocation is a callback implementation details and should be hidden from a caller.

nkumar04 replied:
fixed

toly-kournik commented on bftengine/src/bftengine/DbCheckpointManager.cpp:
```diff
@@ -0,0 +1,235 @@
+Status DbCheckpointManager::createDbCheckpoint(const CheckpointId& checkPointId,
+                                               const uint64_t& lastBlockId,
+                                               const uint64_t& seqNum) {
+  if (!maxNumOfCheckpoints_) return Status::OK();
```
nit: maxNumOfCheckpoints_ == 0

nkumar04 replied:
fixed

toly-kournik commented on bftengine/src/bftengine/DbCheckpointManager.cpp:
```diff
@@ -0,0 +1,235 @@
+    lastCheckpointSeqNum_ = seqNum;
+    dbCheckptMetadata_.dbCheckPoints_.insert(
+        {checkPointId, {checkPointId, lastCheckpointCreationTime_, lastBlockId, lastCheckpointSeqNum_}});
+    while (dbCheckptMetadata_.dbCheckPoints_.size() > maxNumOfCheckpoints_) {
```
Isn't it enough to remove only the first one?

nkumar04 replied:
Yes you are right in most of the cases its enough to remove the first one. 
But, we can have some edge case like, we change configuration for numOfDbCheckpoints. Reduce it from N to N-K, then we may have to remove K db checkpoints

toly-kournik commented on bftengine/src/bftengine/DbCheckpointManager.cpp:
```diff
@@ -0,0 +1,235 @@
+void DbCheckpointManager::checkforCleanup() {
+  std::unique_lock<std::mutex> lk(lock_);
+  while (checkpointToBeRemoved_.empty()) {
+    cv_.wait_for(lk, std::chrono::seconds(2), [this]() { return !checkpointToBeRemoved_.empty(); });
```
why wait for 2 secs and not forever?

nkumar04 replied:
fixed

toly-kournik reviewed (approved):
Few nit comments.
LGTM! Thanks @nkumar04

nkumar04 commented on bftengine/include/bftengine/DbCheckpointManager.hpp:
```diff
@@ -0,0 +1,108 @@
+
+#pragma once
+
+#include "Serializable.h"
```
fixed
